File: Game_Of_Life_Python/Organisms/Animals/human.py
from Organisms.organism import OrganismInitialData
from Organisms.animal import Animal
import pygame


class Human(Animal):

    ABILITY_STRENGTH_INCREASE = 10

    def __init__(self, name, row, column, given_strength=-1, given_initiative=-1, default_strength = None, ability_activated = None):
        strength = OrganismInitialData[name]["strength"]
        initiative = OrganismInitialData[name]["initiative"]
        character = OrganismInitialData[name]["character"]
        if given_strength != -1:
            strength = given_strength
        if given_initiative != -1:
            initiative = given_initiative
        super().__init__(strength, initiative, name, character, row, column, "human.png")
        self.type = Human
        if ability_activated == True:
            self.ability_activated = True
            self.default_strength = default_strength
        else:
            self.ability_activated = False
            self.default_strength = strength

    def action(self, grid_board):

        self.previous_row = self.row
        self.previous_column = self.column

        grid_board[self.row][self.column] = 'e'
        self.print_to_journal(f"{self.name} from ({self.row}, {self.column}) to ")

        running = True
        while running:
            for event in pygame.event.get():
                if event.type == pygame.KEYDOWN:
                    if event.key == pygame.K_LEFT:
                        if self.column > 0:
                            self.organism_go_left()
                        running = False
                    elif event.key == pygame.K_RIGHT:
                        if self.column < len(grid_board[0]) - 1:
                            self.organism_go_right()
                        running = False
                    elif event.key == pygame.K_UP:
                        if self.row > 0:
                            self.organism_go_top()
                        running = False
                    elif event.key == pygame.K_DOWN:
                        if self.row < len(grid_board) - 1:
                            self.organism_go_bottom()
                        running = False

        grid_board[self.row][self.column] = self.character
        self.print_to_journal(f"""({self.row}, {self.column})\n""")

        if self.ability_activated:
            self.strength -= 1
            if self.strength == self.default_strength:
                self.ability_activated = False

    def collision(self, grid_board, organisms, current_index):
        return self.default_collision_animal(grid_board, organisms, current_index)

    def get_static_counter(self):
        return 1

    def increase_static_counter(self):
        pass

    def decrease_static_counter(self):
        pass

    def get_default_strength(self):
        return self.default_strength

    def get_ability_activated(self):
        return self.ability_activated
    
    def set_default_strength(self, value):
        self.strength = value

    def activate_ability(self):
        
        if not self.ability_activated:
            self.ability_activated = True
            self.strength += Human.ABILITY_STRENGTH_INCREASE
            return True
        return False

File: GameOfLife_Cpp/source/src/human.cpp
#include <iostream>
#include "human.hpp"

int Human::HUMAN_STATIC_COUNTER = 0;

Human::Human(int row, int column)
    : Animal(5, 4, 0, "Human", 'H', row, column, "human.png", OrganismType::HUMAN)
{
    this->previous_row = row;
    this->previous_column = column;
    HUMAN_STATIC_COUNTER++;
    ability_activated = false;
    ability_counter = 0;
}

Human::Human(int row, int column, int strength, int initiative, int age, bool activated, int counter, int def_strength)
    : Animal(strength, initiative, age, "Human", 'H', row, column, "human.png", OrganismType::HUMAN)
{
    this->previous_row = row;
    this->previous_column = column;
    HUMAN_STATIC_COUNTER++;
    ability_activated = activated;
    ability_counter = counter;
    this->get_default_strength() = def_strength;
}

Human::Human(const Human& other)
    : Animal(other.get_strength(), other.get_initiative(), other.get_age(), "Human", 'H', other.get_row(), other.get_column(), "human.png", OrganismType::HUMAN)
{
    this->previous_row = other.previous_row;
    this->previous_column = other.previous_column;
    HUMAN_STATIC_COUNTER++;
    ability_activated = other.ability_activated;
    ability_counter = other.ability_counter;
    this->get_default_strength() = other.get_default_strength();
}

Human::~Human()
{
    HUMAN_STATIC_COUNTER--;
}

ActionResult Human::action(std::vector<std::vector<char>>& grid_board)
{
    if (ability_activated)
    {
        this->get_strength()--;
        if (this->get_strength() == this->get_default_strength())
        {
            ability_activated = false;
            ability_counter = 5;
        }
    }
    else if (!ability_activated && ability_counter != 0)
        ability_counter--;

    bool isRunning = true;
    while (isRunning) 
    {
        SDL_Event event;
        while (SDL_PollEvent(&event)) 
        {
            if (event.type == SDL_QUIT) 
            {
                exit(EXIT_SUCCESS);
            }
            else if (event.type == SDL_KEYDOWN)
            {
                bool moved = false;
                this->previous_row = this->row;
                this->previous_column = this->column;
                switch (event.key.keysym.sym) 
                {
                    case SDLK_q:
                        exit(EXIT_SUCCESS);
                        break;
                    case SDLK_a:
                        if (ability_activated)
                            std::cout << "Human ability arleady activated...\n";
                        else if (!ability_activated && ability_counter == 0)
                        {
                            std::cout << "Activating human ability +10 strength\n";
                            ability_activated = true;
                            this->get_default_strength() = this->get_strength();
                            this->get_strength() += 10;
                        }
                        else
                            std::cout << "Cannot activate ability now, you have to wait " << ability_counter << " more turns\n";
                        break;
                    case SDLK_UP:
                        if (this->row > 0)
                        {
                            this->row--;
                            moved = true;
                        }
                        break;
                    case SDLK_DOWN:
                        if (this->row < this->BOARD_HEIGHT - 1)
                        {
                            this->row++;
                            moved = true;
                        }
                        break;
                    case SDLK_LEFT:
                        if (this->column > 0)
                        {
                            this->column--;
                            moved = true;
                        }
                        break;
                    case SDLK_RIGHT:
                        if (this->column < this->BOARD_WIDTH - 1)
                        {
                            this->column++;
                            moved = true;
                        }
                        break;
                }
                if (moved)
                    isRunning = false;
            }
        }
    }
    
    this->move_message();
    grid_board[previous_row][previous_column] = 'e';
    grid_board[row][column] = this->get_character();

    return ActionResult(ActionType::MOVE);
}

CollisionResult Human::collision(std::vector<std::vector<char>>& grid_board, std::vector<Organism*>& organisms, int current_index)
{
    return this->default_collision_animal(grid_board, organisms, current_index);
}

int Human::get_static_counter()
{
    return HUMAN_STATIC_COUNTER;
}